### PR TITLE
feat(enforcement): auto-fix bare python calls by prepending uv run

### PR DESCRIPTION
**Pull request by myakove**
Instead of blocking bare `python`/`python3` commands, auto-fix them by prepending `uv run`.

## Changes

- **enforcement-helpers.ts**: `checkPythonPipBlock` now returns `autofix` for python/python3 (prepends `uv run`) and `block` for pip/pip3. Added `setUvAvailable`/`isUvAvailable` state.
- **enforcement.ts**: Handle `autofix` result by mutating `event.input.command` instead of blocking.
- **session-validation.ts**: Check uv availability at session start, offer to install if missing, call `setUvAvailable()`. When uv is not installed, all python/pip enforcement is disabled.
- **tests**: 23 tests covering auto-fix, pip block, uv-unavailable passthrough, compound commands, env vars, path-prefixed calls.

## Behavior

| Command | uv available | Result |
|---------|-------------|--------|
| `python3 -c 'code'` | ✅ | Auto-fixed → `uv run python3 -c 'code'` |
| `/usr/bin/python3 script.py` | ✅ | Auto-fixed → `uv run python3 script.py` |
| `LANG=C python3 foo` | ✅ | Auto-fixed → `LANG=C uv run python3 foo` |
| `ls && python3 -c 'x'` | ✅ | Auto-fixed → `ls && uv run python3 -c 'x'` |
| `pip install requests` | ✅ | Blocked (use `uv add`) |
| `python3 anything` | ❌ | Allowed through (no enforcement) |
| `pip install anything` | ❌ | Allowed through (no enforcement) |

Made with [Cursor](https://cursor.com)

### CODE REVIEW

File: extensions/orchestrator/enforcement-helpers.ts [hunk 1]
```diff
@@ -8,7 +8,12 @@ import * as path from "node:path";
 import { join } from "node:path";
 import { DANGEROUS, hasGitSub } from "./git-helpers.js";
 
-export type EnforcementResult = { block: true; reason: string } | undefined;
+export type EnforcementResult = { block: true; reason: string } | { autofix: true; modifiedCommand: string; reason: string } | undefined;
+
+/** Whether uv is available on this system (checked at session_start) */
+let uvAvailable = true;
+export function setUvAvailable(val: boolean): void { uvAvailable = val; }
+export function isUvAvailable(): boolean { return uvAvailable; }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts:16]: Fixed: added direct test for isUvAvailable() getter that verifies get/set contract.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 2]
```diff
@@ -50,23 +55,54 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+        const modifiedCommand = command.replace(origStmt, fixedStmt);
+
         return {
-          block: true,
-          reason:
-            "Direct python/pip forbidden. Use: uv run python3 / uv run script.py / uvx tool / uv add pkg",
+          autofix: true,
+          modifiedCommand,
+          reason: "Auto-fixed: prepended `uv run` to python command",
         };
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts:101]: Fixed: updated README to reflect auto-fix behavior instead of block.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 3]
```diff
@@ -50,23 +55,54 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+
+        // python/python3: auto-fix by prepending `uv run` to the python call
+        // Find where python appears in the original statement (after env vars, with path)
+        const origStmt = statementsOrig[i] || stmtLower;
+        const envVarMatch = origStmt.match(envVarPrefixRe);
+        const envPrefix = envVarMatch?.[0] || "";
+        const afterEnv = origStmt.slice(envPrefix.length);
+        // Replace the python command (possibly path-prefixed) with `uv run python3`
+        const fixedAfterEnv = afterEnv.replace(/^(\S*\/)?python3?\b/, "uv run python3");
+        const fixedStmt = envPrefix + fixedAfterEnv;
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: replacement now uses `uv run ${baseCmd}` to preserve the original interpreter name (python vs python3).

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 4]
```diff
@@ -50,23 +55,54 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+
+        // Reconstruct full command by replacing the original statement
+        // We need to find and replace the exact original statement in the full command
+        const modifiedCommand = command.replace(origStmt, fixedStmt);
+
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: replaced substring-based command.replace() with offset-based slicing using matchAll to track segment positions. Added test for quoted-argument edge case.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/session-validation.ts [hunk 5]
```diff
@@ -191,15 +192,35 @@ function registerRepairCommand(pi: ExtensionAPI): void {
...
 async function checkSessionTools(ctx: any): Promise<void> {
+  // ── uv availability check ────────────────────────────────────────
+  const uvInstalled = hasCmd("uv");
+  setUvAvailable(uvInstalled);
+  if (!uvInstalled && ctx.hasUI) {
+    try {
+      const wantInstall = await ctx.ui.confirm(
+        "⚡ uv is not installed",
+        "uv enables faster Python package management and auto-fix for python/pip commands.\nInstall it now? (curl -LsSf https://astral.sh/uv/install.sh | sh)",
+      );
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/session-validation.ts]: Fixed: moved uv availability check before the `if (!ctx.hasUI) return` guard. Headless sessions now correctly detect uv.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/session-validation.ts [hunk 6]
```diff
@@ -191,15 +192,35 @@ function registerRepairCommand(pi: ExtensionAPI): void {
...
+        const result = spawnSync("sh", ["-c", "curl -LsSf https://astral.sh/uv/install.sh | sh"], {
+          timeout: 60_000,
+          stdio: ["pipe", "pipe", "pipe"],
+        });
+        if (result.status === 0 && hasCmd("uv")) {
+          setUvAvailable(true);
+          ctx.ui.notify("✅ uv installed successfully", "info");
+        } else {
+          ctx.ui.notify("❌ uv installation failed. Install manually: https://docs.astral.sh/uv/", "warning");
+        }
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/session-validation.ts]: Fixed: removed curl|sh installer. Now shows notify-only message with install instructions. No code execution.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement-helpers.ts:13` (qodo finding) — isUvAvailable() lacks test
> Addressed: Fixed: added direct test for isUvAvailable() getter that verifies get/set contract.
### `extensions/orchestrator/session-validation.ts:195` (qodo bug) — Headless uv availability wrong
> Addressed: Fixed: moved uv availability check before the `if (!ctx.hasUI) return` guard. Headless sessions now correctly detect uv.
### `extensions/orchestrator/enforcement-helpers.ts:88` (qodo bug) — Python rewritten to python3
> Addressed: Fixed: replacement now uses `uv run ${baseCmd}` to preserve the original interpreter name (python vs python3).
### `extensions/orchestrator/enforcement-helpers.ts:58` (qodo rule violation) — README python enforcement outdated
> Addressed: Fixed: updated README to reflect auto-fix behavior instead of block.
### `extensions/orchestrator/enforcement-helpers.ts:68` (qodo bug) — Autofix replaces wrong segment
> Addressed: Fixed: replaced substring-based command.replace() with offset-based slicing using matchAll to track segment positions. Added test for quoted-argument edge case.
### `extensions/orchestrator/session-validation.ts:200` (qodo bug) — curl|sh installer execution
> Addressed: Fixed: removed curl|sh installer. Now shows notify-only message with install instructions. No code execution.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: tests/node/orchestrator/enforcement.test.ts [hunk 7]
```diff
@@ -439,66 +441,137 @@ describe("resolveEffectiveCwd", () => {
...
+    const r = checkPythonPipBlock("python3 --version", "python3 --version");
+    assert.ok(r && "autofix" in r);
+    assert.equal(r.modifiedCommand, "uv run python3 --version");
+  });
+  it("auto-fixes python script.py", () => {
+    const r = checkPythonPipBlock("python script.py", "python script.py");
+    assert.ok(r && "autofix" in r);
+    assert.equal(r.modifiedCommand, "uv run python script.py");
+  });
+  it("auto-fixes real python when same text appears earlier in quotes", () => {
```
[the remark this answers is not in the document]
Comment (myakove) [reply tests/node/orchestrator/enforcement.test.ts]: Fixed: renamed test to 'auto-fixes correct segment with python in earlier quotes' — removed 'and'.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 8]
```diff
@@ -50,23 +55,70 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+        const afterEnv = origText.slice(envPrefix.length);
+        const fixedAfterEnv = afterEnv.replace(/^(\S*\/)?python3?\b/, `uv run ${baseCmd}`);
+        const fixedStmt = envPrefix + fixedAfterEnv;
+
+        // Replace by offset — find the original untrimmed segment in the command
+        const rawSegment = command.slice(seg.start, seg.end);
+        const trimStart = rawSegment.indexOf(seg.text);
+        const absStart = seg.start + (trimStart >= 0 ? trimStart : 0);
+        const absEnd = absStart + seg.text.length;
+        const modifiedCommand = command.slice(0, absStart) + fixedStmt + command.slice(absEnd);
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: two-pass approach — first pass blocks if ANY segment is pip/pip3, second pass rewrites ALL python segments. Added tests for python+pip compound commands.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 9]
```diff
@@ -50,23 +55,70 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+            reason: "Direct pip/pip3 forbidden. Use: uv add <pkg> / uvx <tool> / uv run --with <pkg> script.py",
+          };
+        }
+
+        // Auto-fix: use the original segment text (preserving case)
+        const origText = seg.text;
+        const envVarMatch = origText.match(envVarPrefixRe);
+        const envPrefix = envVarMatch?.[0] || "";
+        const afterEnv = origText.slice(envPrefix.length);
+        const fixedAfterEnv = afterEnv.replace(/^(\S*\/)?python3?\b/, `uv run ${baseCmd}`);
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: regex is now case-insensitive (`/i` flag) to handle any casing in the original command.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement-helpers.ts:111` (qodo bug) — Pip block bypass
> Addressed: Fixed: two-pass approach — first pass blocks if ANY segment is pip/pip3, second pass rewrites ALL python segments. Added tests for python+pip compound commands.
### `tests/node/orchestrator/enforcement.test.ts:458` (qodo finding) — Test name contains and
> Addressed: Fixed: renamed test to 'auto-fixes correct segment with python in earlier quotes' — removed 'and'.
### `extensions/orchestrator/enforcement-helpers.ts:107` (qodo bug) — No-op python autofix
> Addressed: Fixed: regex is now case-insensitive (`/i` flag) to handle any casing in the original command.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 10]
```diff
@@ -50,26 +55,86 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+
+    // Pass 2: rewrite ALL python/python3 segments
+    let modifiedCommand = command;
+    let anyRewrite = false;
+    // Process segments in reverse order so offsets remain valid after each splice
+    for (let i = segments.length - 1; i >= 0; i--) {
+      const seg = segments[i];
+      if (!seg.textLower) continue;
+      const strippedLower = seg.textLower.replace(envVarPrefixRe, "");
+      const baseCmd = strippedLower.split(/\s/)[0]?.replace(/^.*\//, "");
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: path stripping now handles both `/` and `\\` separators. Added tests for Windows-style `C:\\Python\\python3` and `C:\\Python\\Scripts\\pip` paths.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement-helpers.ts:109` (qodo bug) — Windows path bypass
> Addressed: Fixed: path stripping now handles both `/` and `\\` separators. Added tests for Windows-style `C:\\Python\\python3` and `C:\\Python\\Scripts\\pip` paths.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 11]
```diff
@@ -50,26 +55,86 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+      const strippedLower = seg.textLower.replace(envVarPrefixRe, "");
+      const baseCmd = strippedLower.split(/\s/)[0]?.replace(/^.*[\/\\]/, "");
+
+      if (baseCmd && /^python3?$/.test(baseCmd)) {
+        const origText = seg.text;
+        const envVarMatch = origText.match(envVarPrefixRe);
+        const envPrefix = envVarMatch?.[0] || "";
+        const afterEnv = origText.slice(envPrefix.length);
+        // Case-insensitive regex to handle any casing
+        const fixedAfterEnv = afterEnv.replace(/^(\S*[\/\\])?python3?\b/i, `uv run ${baseCmd}`);
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: now extracts original-cased executable from the original segment text via `afterEnv.match()` instead of using lowercased `baseCmd`. Added test: `Python3 --version` → `uv run Python3 --version`.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement-helpers.ts:109` (qodo bug) — Autofix lowercases executable
> Addressed: Fixed: now extracts original-cased executable from the original segment text via `afterEnv.match()` instead of using lowercased `baseCmd`. Added test: `Python3 --version` → `uv run Python3 --version`.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

File: extensions/orchestrator/enforcement-helpers.ts [hunk 12]
```diff
@@ -50,26 +55,88 @@ export function resolveEffectiveCwd(command: string, sessionCwd: string): string
...
+
+      if (baseCmd && /^python3?$/.test(baseCmd)) {
+        const origText = seg.text;
+        const envVarMatch = origText.match(envVarPrefixRe);
+        const envPrefix = envVarMatch?.[0] || "";
+        const afterEnv = origText.slice(envPrefix.length);
+        // Extract original-cased executable name, then prepend uv run
+        const execMatch = afterEnv.match(/^(\S*[\/\\])?(python3?)\b/i);
+        const origExe = execMatch?.[2] || baseCmd;
+        const fixedAfterEnv = afterEnv.replace(/^(\S*[\/\\])?python3?\b/i, `uv run ${origExe}`);
```
[the remark this answers is not in the document]
Comment (myakove) [reply extensions/orchestrator/enforcement-helpers.ts]: Fixed: first-token extraction is now quote-aware — handles both `"path with spaces/python3"` and `'path/python3'`. Pass 2 rewrite strips quoted paths correctly. Added tests for quoted Windows paths with spaces.

---
*Assisted-by: PI (claude-opus-4-6-1m)*

**Comment by myakove** [comment]
/qodo

The following review comments were reviewed and a decision was made:

### `extensions/orchestrator/enforcement-helpers.ts:117` (qodo bug) — Quoted python path bypass
> Addressed: Fixed: first-token extraction is now quote-aware — handles both `"path with spaces/python3"` and `'path/python3'`. Pass 2 rewrite strips quoted paths correctly. Added tests for quoted Windows paths with spaces.

---
*Assisted-by: PI (claude-opus-4-6-1m)*
